File: src/Layout/Areas/Layouter.java

package Layout.Areas;

import java.util.Objects;

import javafx.animation.*;
import javafx.event.EventHandler;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.Pane;

import Layout.container.Container;
import Layout.container.bicontainer.BiContainer;
import Layout.container.freeformcontainer.FreeFormContainer;
import gui.GUI;
import gui.objects.Pickers.Picker;
import gui.objects.Pickers.WidgetPicker;
import util.animation.interpolator.CircularInterpolator;
import util.graphics.drag.DragUtil;

import static de.jensd.fx.glyphs.fontawesome.FontAwesomeIcon.EXCHANGE;
import static gui.GUI.*;
import static javafx.animation.Interpolator.LINEAR;
import static javafx.geometry.Orientation.HORIZONTAL;
import static javafx.geometry.Orientation.VERTICAL;
import static javafx.scene.input.MouseButton.PRIMARY;
import static javafx.scene.input.MouseEvent.MOUSE_CLICKED;
import static javafx.scene.input.MouseEvent.MOUSE_ENTERED;
import static main.App.APP;
import static util.animation.interpolator.EasingMode.EASE_OUT;
import static util.functional.Util.stream;
import static util.graphics.Util.setAnchor;
import static util.graphics.Util.setAnchors;
import static util.graphics.drag.DragUtil.installDrag;

/**
 * @author uranium
 *
 * @TODO make dynamic indexes work and this widget part of layout map. See
 * TO Do file API section
 */
@   Layout.widget.Widget.Info
public final class Layouter implements ContainerNode {

    private final Container container;
    private final int index;

    public final Picker<String> cp = new Picker();
    public final AnchorPane root = new AnchorPane();

    private final FadeTransition a1;
    private final ScaleTransition a2;
    private final EventHandler<MouseEvent> clickShowHider;
    private final EventHandler<MouseEvent> exitHider;

    public Layouter(Container con, int index) {
        Objects.requireNonNull(con);

        this.index = index;
        this.container = con;

        cp.onSelect = layout -> {
            switch(layout) {
                case "Split Vertically" : closeAndDo(cp.root, this::showSplitV);
                                        break;
                case "Split Horizontally" : closeAndDo(cp.root, this::showSplitH);
                                        break;
                case "Widget" : closeAndDo(cp.root, this::showWidgetArea);
                                break;
                case "FreeForm" : closeAndDo(cp.root, this::showFreeform);
                                  break;
            }
        };
        cp.onCancel = this::hide;
        cp.textCoverter = text -> text;
        cp.itemSupply = () -> stream("Split Vertically", "Split Horizontally",
                                        "Widget", "FreeForm"); // , "Tabs"
        cp.buildContent();
        setAnchor(root, cp.root,0d);

        Interpolator i = new CircularInterpolator(EASE_OUT);
        a1 = new FadeTransition(ANIM_DUR, cp.root);
        a1.setInterpolator(LINEAR);
        a2 = new ScaleTransition(ANIM_DUR, cp.root);
        a2.setInterpolator(i);

        cp.root.setOpacity(0);
        cp.root.setScaleX(0);
        cp.root.setScaleY(0);

        // drag&drop
        installDrag(
            root, EXCHANGE, "Switch components",
            DragUtil::hasComponent,
            e -> container == DragUtil.getComponent(e),
            e -> DragUtil.getComponent(e).swapWith(container,index)
        );

        clickShowHider =  e -> {
            if(e.getButton()==PRIMARY) {
                if(cp.root.getOpacity()!=0) return;
                // avoid when under lock
                if(container.lockedUnder.get()) return;
                // rely on the public show() implementation, not internal one
                show();
                e.consume();
            }

        };
//        exitHider =  e -> cp.onCancel.run();
        exitHider = e -> {
            if(cp.root.getScaleX()==1 && wp.root.getScaleX()==1) {
                // rely on the public show() implementation, not internal one
                cp.onCancel.run();
                e.consume();
            }
        };

        // setParentRec mode
        setWeakMode(true); // this needs to be called in constructor
        // setParentRec show
        setShow(GUI.isLayoutMode());
    }

/****************************  functionality  *********************************/

    @Override
    public void show() {
        showControls(true);
//        openAndDo(cp.root, null);
    }

    @Override
    public void hide() {
        // prevent leaving layout mode when layout mode active
        if(GUI.isLayoutMode())return;
        showControls(false);
//        closeAndDo(cp.root, null);
    }

    private void showControls(boolean val) {
        a1.stop();
        a2.stop();
        if (val) {
            a1.setToValue(1);
            a2.setToX(1);
            a2.setToY(1);
        } else {
            a1.setToValue(0);
            a2.setToX(0);
            a2.setToY(0);
        }
        a1.play();
        a2.play();
    }

    private boolean clickMode = true;

    /**
     * In normal mode the controls are displayed on mouse click
     * In weak mode the controls are displayed on mouse hover
     * Default false.
     * @param val
     */
    public void setWeakMode(boolean val) {
        clickMode = val;

        // always hide on mouse exit, setParentRec
        if (root.getOnMouseExited()==null)
            root.setOnMouseExited(exitHider);
        // swap handlers
        if(val) {
            root.addEventHandler(MOUSE_CLICKED,clickShowHider);
            root.removeEventHandler(MOUSE_ENTERED,clickShowHider);
        } else {
            root.addEventHandler(MOUSE_CLICKED,clickShowHider);
            root.removeEventHandler(MOUSE_ENTERED,clickShowHider);
        }
    }

    public void toggleWeakMode() {
        clickMode = !clickMode;
    }
    public boolean isWeakMode() {
        return clickMode;
    }


    WidgetPicker wp = new WidgetPicker();
    private void showWidgetArea() {
        wp = new WidgetPicker();
        wp.onSelect = factory -> {
            closeAndDo(wp.root, () -> {
                root.getChildren().remove(wp.root);
                root.setOnMouseExited(null);
                // this is the crucial part
                container.addChild(index, factory.create());
                if(GUI.isLayoutMode()) container.show();
                APP.actionStream.push("New widget");
            });
        };
        wp.onCancel = () -> closeAndDo(wp.root, () -> {
            root.getChildren().remove(wp.root);
            showControls(true);
        });
        wp.consumeCancelClick = true;
        wp.buildContent();
        root.getChildren().add(wp.root);
        setAnchors(wp.root, 0d);
        openAndDo(wp.root, null);
    }
    private void showSplitV() {
        container.addChild(index, new BiContainer(HORIZONTAL));
        APP.actionStream.push("Divide layout");
    }
    private void showSplitH() {
        container.addChild(index, new BiContainer(VERTICAL));
        APP.actionStream.push("Divide layout");
    }
    private void showFreeform() {
        container.addChild(index, new FreeFormContainer());
    }

    @Override
    public Pane getRoot() {
        return root;
    }
}

File: src/player/main/sp/it/pl/layout/Layouter.kt
package sp.it.pl.layout

import javafx.scene.Node
import javafx.scene.input.MouseButton.PRIMARY
import javafx.scene.input.MouseEvent.MOUSE_CLICKED
import javafx.scene.input.MouseEvent.MOUSE_EXITED
import javafx.scene.layout.Pane
import javafx.scene.layout.StackPane
import sp.it.pl.main.APP
import sp.it.pl.main.AppAnimator
import sp.it.pl.main.Df
import sp.it.pl.main.IconFA
import sp.it.pl.main.contains
import sp.it.pl.main.get
import sp.it.pl.main.installDrag
import sp.it.pl.ui.objects.picker.ContainerPicker
import sp.it.pl.ui.objects.picker.Picker
import sp.it.pl.ui.objects.picker.WidgetPicker
import sp.it.util.async.coroutine.FX
import sp.it.util.async.coroutine.launch
import sp.it.util.reactive.Disposer
import sp.it.util.reactive.attach
import sp.it.util.reactive.on
import sp.it.util.reactive.onEventDown
import sp.it.util.ui.containsMouse
import sp.it.util.ui.lay

/**
 * Container graphics for container null child.
 *
 * Shows user component picker to populate container with content (at this index) - it allows creation of layouts. Uses
 * two nested [Picker]s, one for [Container] and the other for widgets/templates.
 */
class Layouter: ComponentUi {

   private val container: Container<*>
   private val index: Int

   override val root = StackPane()
   var onCancel: () -> Unit = {}
   private var wasSelected = false
   private val cp: ContainerPicker
   private var isCpShown = false
   private val disposer = Disposer()

   @JvmOverloads
   constructor(container: Container<*>, index: Int, exitRoot: Pane? = null) {
      this.container = container
      this.index = index
      this.cp = ContainerPicker({ showContainer(it) }, { showWidgetArea(it) }).apply {
         onSelect = {
            wasSelected = true
            isCpShown = false
            AppAnimator.closeAndDo(root) { it.onSelect() }
         }
         onCancel = {
            if (!APP.ui.isLayoutMode)
               hide()
         }
         consumeCancelEvent = false
         buildContent()
      }
      APP.widgetManager.widgets.separateWidgets attach { cp.buildContent() } on disposer

      root.lay += cp.root

      AppAnimator.applyAt(cp.root, 0.0)

      root.installDrag(
         IconFA.EXCHANGE,
         "Switch components",
         { e -> Df.COMPONENT in e.dragboard },
         { e -> e.dragboard[Df.COMPONENT]===container },
         { e -> e.dragboard[Df.COMPONENT].swapWith(container, index) }
      )

      // show cp on mouse click
      root.onEventDown(MOUSE_CLICKED, PRIMARY, false) {
         if (!container.lockedUnder.value) {
            show()
            it.consume()
         }
      }

      // hide cp on mouse exit
      val er = exitRoot ?: cp.root
      er.onEventDown(MOUSE_EXITED) {
         if (!wasSelected && !isSuppressedHiding(er)) {
            cp.onCancel()
            it.consume()
         }
      } on disposer
   }

   override fun show() = showControls(true)

   override fun hide() = showControls(false)

   override fun dispose() = disposer()

   fun hideAnd(onHidden: () -> Unit) = showControls(false) { onHidden() }

   private fun showControls(value: Boolean, onClosed: () -> Unit = {}) {
      if (isCpShown==value) {
         if (!value) onClosed()
         return
      }
      val isWpShown = root.children.size!=1
      if (isWpShown) {
         if (!value) onClosed()
         return
      }

      isCpShown = value
      if (value) {
         AppAnimator.openAndDo(cp.root, null)
      } else {
         AppAnimator.closeAndDo(cp.root) {
            if (!wasSelected) onCancel()
            onClosed()
         }
      }
   }

   private fun showContainer(c: Container<*>) {
      container.addChild(index, c)
      if (c is ContainerBi) APP.actionStream("Divide layout")
   }

   private fun showWidgetArea(mode: WidgetPicker.Mode) {
      val wp = WidgetPicker(mode)
      wp.onSelect = { factory ->
         wasSelected = false
         AppAnimator.closeAndDo(wp.root) {
            root.children -= wp.root
            launch(FX) {
               container.addChild(index, factory.create())
               if (APP.ui.isLayoutMode) container.show()
               APP.actionStream("New widget")
            }
         }
      }
      wp.onCancel = {
         wasSelected = false
         AppAnimator.closeAndDo(wp.root) {
            root.children -= wp.root
            if (root.containsMouse())
               showControls(true)
         }
      }
      wp.buildContent()
      wp.consumeCancelEvent = true // we need right-click to not close container

      root.lay += wp.root
      AppAnimator.openAndDo(wp.root, null)
   }

   companion object {
      private const val SUPPRESS_HIDING_KEY = "suppress"

      fun suppressHidingFor(node: Node, suppress: Boolean) {
         if (suppress) node.properties[SUPPRESS_HIDING_KEY] = null
         else node.properties -= SUPPRESS_HIDING_KEY
      }

      private fun isSuppressedHiding(node: Node) = SUPPRESS_HIDING_KEY in node.properties
   }
}